torch/utils/checkpoint.py

```diff
@@ -614,21 +609,22 @@ def pack_hook(x):
                 return x.detach()
             holder = target_frame.weak_holders[recomp_idx]()
 
+            # This holder may have been cleared because someone may have called
+            # backward within forward. If so, we don't need to save.
             if holder is not None:
-                # See Rule 6: [ Multiple backwards ] above
-                if holder.handles.get(gid, None) is None:
-                    holder.handles[gid] = _Handle()
+                assert holder.handles.get(gid, None) is None, "this is a bug, please file an issue"
+                holder.handles[gid] = _Handle()
```

Comment:
We can make this assertion because (ignoring any multithreading issues), it is not possible to compute for the same holder twice in the same backward, since after we recompute for the first time, we would've set is_recomputed=True.